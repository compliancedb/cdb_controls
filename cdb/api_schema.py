class ApiSchema:
    @staticmethod
    def url_for_owner_projects(host, project_data):
        return host + '/api/v1/projects/' + project_data["owner"] + "/"

    @staticmethod
    def url_for_project(host, project_data):
        return ApiSchema.url_for_owner_projects(host, project_data) + project_data["name"]

    @staticmethod
    def url_for_releases(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/releases/'

    @staticmethod
    def url_for_commit(host, project_data, commit):
        return ApiSchema.url_for_project(host, project_data) + '/commits/' + commit

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_artifact(host, project_data, sha256_digest):
        return ApiSchema.url_for_artifacts(host, project_data) + sha256_digest

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_release(host, project_data, release_number):
        return ApiSchema.url_for_releases(host, project_data) + release_number