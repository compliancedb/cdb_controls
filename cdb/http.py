import json
import os

import requests as req
from requests.auth import HTTPBasicAuth
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry


def http_get_json(url, api_token):
    print("Getting this endpoint: " + url)
    resp = req.get(url, auth=HTTPBasicAuth(api_token, 'unused'))
    print(resp.text)
    return resp.json()


def put_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Putting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    if os.getenv('CDB_DRY_RUN', "FALSE") != "TRUE":
        resp = req.put(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
        print(resp.text)
    else:
        print("DRY RUN: Put not sent")


def http_post_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Posting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    if os.getenv('CDB_DRY_RUN', "FALSE") != "TRUE":
        resp = retry_http().post(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
        print(resp.text)
    else:
        print("DRY RUN: POST not sent")


def retry_http():
    # The backoff_factor controls how long to sleep between retries. The algorithm is
    # {backoff factor} * (2 ** ({number of total retries} - 1))
    # So for 1, the 5 retry sleeps are [ 0.5, 1, 2, 4, 8 ]
    # for a total sleep of 15.5 seconds
    strategy = Retry(
        total=5,
        backoff_factor=1,
        status_forcelist=[503],
        allowed_methods=["POST"]
    )
    adapter = HTTPAdapter(max_retries=strategy)
    s = req.Session()
    s.mount("https://", adapter)
    s.mount("http://", adapter)
    return s
