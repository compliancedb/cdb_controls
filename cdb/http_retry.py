from requests import Session
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry
from os import sys
from time import sleep

import requests as req

MAX_RETRY_COUNT = 5
RETRY_BACKOFF_FACTOR = 1


def http_retry_get(url, auth):
    return HttpRetry().get(url, auth)


class HttpRetry():
    MAX_RETRY_COUNT = 5

    def get(self, url, auth):
        response = req.get(url, auth=auth)
        if response.status_code != 503:
            return response

        err_print("The HTTP call failed. Retrying...")
        for count in range(1, MAX_RETRY_COUNT + 1):
            sleep(self._sleep_time(count))
            response = req.get(url, auth=auth)
            if response.status_code != 503:
                return response
            else:
                self._log_retry_failure(count, response)

        raise req.exceptions.RetryError("TODO")

    def _log_retry_failure(self, count, response):
        err_print("Retry {}/{} failed, status={}{}".format(
            count,
            MAX_RETRY_COUNT,
            response.status_code,
            self._sleep_message(count)
        ))

    def _sleep_message(self, count):
        if count < MAX_RETRY_COUNT:
            return ", sleeping for {} seconds...".format(self._sleep_time(count))
        else:
            return ""

    def _sleep_time(self, count):
        return RETRY_BACKOFF_FACTOR * (2 ** count)



def http_retry():
    """
    Create a http requests object with an embedded retry strategy.
    Affects only GET,PUT,POST requests whose response status is 503.
    Intention is to hide small server downtimes during deployments.
    The sleep duration between retries is [0,2,4,8,16] seconds.
    Retries 5 times. The first retry is immediate.

    The logged message does _not_ say 'Press Control^C to exit.'
    because Control^C requires a runtime environment with a tty.
    That would require a [docker run -it] which would fail in the
    target environment, a CI pipeline, which invariably has no tty.
    """
    strategy = LoggingRetry(
        total=MAX_RETRY_COUNT,
        backoff_factor=RETRY_BACKOFF_FACTOR,
        status_forcelist=[503],
        allowed_methods=["GET", "POST", "PUT"]
    )
    adapter = HTTPAdapter(max_retries=strategy)
    session = Session()
    session.mount("https://", adapter)
    session.mount("http://", adapter)
    return session


class LoggingRetry(Retry):
    """
    Superclass source and docs are here:
    https://github.com/urllib3/urllib3/blob/master/src/urllib3/util/retry.py
    https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry

    I tried moving the kwargs in http_retry() into a LoggingRetry.__init__(...)
    I got: TypeError: __init__() got an unexpected keyword argument
    Don't know why, so for now they stay in http_retry()

    Adding attributes to this class causes errors due, I think, to the Retry
    super-class doing clever things with __init__(), new(), and __getattr__()
    """
    def increment(self, *args, **kwargs):
        if self.count() == 0:
            err_print("The HTTP call failed. Retrying...")
        new_retry = super().increment(*args, **kwargs)
        new_retry.log_increment()
        return new_retry

    def log_increment(self):
        err_print("Retry {}/{} failed, status={}{}".format(
            self.count(),
            MAX_RETRY_COUNT,
            self.history[-1].status,
            self.sleep_message()
        ))

    def count(self):
        return len(self.history)

    def sleep_message(self):
        if self.count() < MAX_RETRY_COUNT:
            return ", sleeping for {} seconds...".format(self.sleep_time())
        else:
            return ""

    def sleep_time(self, n=None):
        """"
        Retry documentation says the backoff algorithm is:
           {backoff factor} * (2 ** ({number of retries} - 1))
        With backoff_factor==1, this would give successive sleeps of:
           [ 0.5, 1, 2, 4, 8, 16, ... ]
        However, the _actual_ sleep durations you get from get_backoff_time() are:
           [ 0, 2, 4, 8, 16, ... ]
        And empirically, before the first retry there is indeed no sleep.
        So the first retry is happening immediately.
        """
        if n is None:
            n = self.count()
        if n == 0:
            return 0
        else:
            return RETRY_BACKOFF_FACTOR * (2 ** n)


def total_sleep_time():
    return sum(LoggingRetry().sleep_time(n) for n in range(0, MAX_RETRY_COUNT))


def err_print(message):
    # flush is handy for manual test run with capsys off
    print(message, file=sys.stderr, flush=True)
