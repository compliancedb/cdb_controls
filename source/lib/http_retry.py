"""
Wraps http requests with a simple retry strategy.
Affects only GET,PUT,POST requests whose response status is 503.
Intention is to hide small server downtimes during deployments.
Retries 5 times with successive sleep durations of [1,2,4,8,16] seconds.

The logged message does _not_ say 'Press Control^C to exit.'
because Control^C requires a runtime environment with a tty.
That would require a [docker run -it] which would fail in the
target environment, a CI pipeline, which invariably has no tty.

Originally we implemented http retries with this Retry class:
https://github.com/urllib3/urllib3/blob/master/src/urllib3/util/retry.py
https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
We soon abandoned it.
- The only way we could mock the deeply occurring http calls was with httpretty
  which mocks at the socket level and has no way to mock only the
  first N calls in a retry situation. Thus it was impossible to test
  the happy-path scenario of a few 503's followed by a 200/201.
- It proved impossible create a Retry subclass containing attributes because
  of clever meta-programming in its __init__, new, and __getattr__
- Its empirical behaviour differed from its limited documented behaviour.
  Eg, its sleep backoff actually had _no_ sleep before the first retry.
- Its implementation was hard to understand.
  Eg, its crucial increment() method returned a _new_ Retry() object.
- In short, it was complicated and hard to test.
"""

from errors import ChangeError
import requests as http
from time import sleep

RETRY_BACKOFF_FACTOR = 1
MAX_RETRY_COUNT = 5


class HttpRetryExhausted(ChangeError):
    def __init__(self, url):
        self._url = url

    def url(self):
        return self._url


class HttpRetry():
    def __init__(self, stdout):
        self._status_retry_list = [503]
        self._max_count = MAX_RETRY_COUNT
        self._stdout = stdout

    def total_sleep_time(self):
        return sum(self._sleep_time(n) for n in range(0, self._max_count))

    def get(self, url, **kwargs):
        return self._http_retry(url, lambda: http.get(url, **kwargs))

    def put(self, url, **kwargs):
        return self._http_retry(url, lambda: http.put(url, **kwargs))

    def post(self, url, **kwargs):
        return self._http_retry(url, lambda: http.post(url, **kwargs))

    def _http_retry(self, url, http_call):
        response = http_call()
        status = response.status_code
        if not self._retry(status):
            return response
        count = 0
        seconds = self._sleep_time(count)
        self._log(count, status, seconds)
        for count in range(1, self._max_count + 1):
            seconds = self._sleep_time(count)
            sleep(seconds)
            response = http_call()
            status = response.status_code
            self._log(count, status, seconds)
            if not self._retry(status):
                return response

        raise HttpRetryExhausted(url)

    def _retry(self, status):
        return status in self._status_retry_list

    def _log(self, count, status, seconds):
        lhs = self._response_message(count, status)
        rhs = self._retry_message(count, status, seconds)
        self._stdout.print(f"{lhs}{rhs}")

    def _response_message(self, count, status):
        if count == 0:
            return f"Response.status={status}"
        else:
            return f"Retry {count}/{self._max_count}: response.status={status}"

    def _retry_message(self, count, status, seconds):
        if not self._retry(status):
            return ""
        elif count < self._max_count:
            return f", retrying in {seconds} seconds..."
        else:
            return ""

    @staticmethod
    def _sleep_time(count):
        return RETRY_BACKOFF_FACTOR * (2 ** count)

