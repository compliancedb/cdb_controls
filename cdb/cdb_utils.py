import getopt
import json
import os
import sys
import docker
import requests as req
from junitparser import JUnitXml
from requests.auth import HTTPBasicAuth

from cdb.settings import CDB_SERVER

CMD_HELP = 'ensure_project.py -p <project.json>'


def project_exists_in_cdb(project_data, projects):
    exists = False
    for project in projects:
        if project["name"] == project_data["name"]:
            exists = True
    return exists


def get_project_list_from_cdb(projects_url):
    resp = req.get(projects_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    projects = resp.json()
    return projects


def get_project_from_cdb(project_url):
    resp = req.get(project_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    project = resp.json()
    return project


def load_project_configuration(json_data_file):
    project_data = json.load(json_data_file)
    return project_data


def create_artifact(api_token, host, project_config_file, sha256, filename, description, git_commit, commit_url, build_url,
                    is_compliant):
    project_data = load_project_configuration(project_config_file)

    '''
    curl -H 'Content-Type: application/json' \
     -X PUT \
     -d '{"sha256": "'"$3"'", "filename": "'"$4"'", "description": "'"$5"'", "git_commit": "'"$6"'", "commit_url": "'"$7"'", "build_url": "'"$8"'", "is_compliant": "true"}' \
    http://server:8001/api/v1/projects/$1/$2/artifacts/
    '''

    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = url_for_artifacts(host, project_data)
    put_payload(create_artifact_payload, url, api_token)


def put_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Putting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    resp = req.put(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
    print(resp.text)


def url_for_artifacts(host, project_data):
    return url_for_project(host, project_data) + project_data["name"] + '/artifacts/'


def add_evidence(api_token, host, project_file_contents, sha256_digest, evidence):
    project_data = load_project_configuration(project_file_contents)
    url_for_artifact = url_for_artifacts(host, project_data) + sha256_digest

    put_payload(evidence, url_for_artifact, api_token)


def url_for_project(host, project_data):
    return host + '/api/v1/projects/' + project_data["owner"] + '/'


def rchop(thestring, ending):
    if thestring.endswith(ending):
        return thestring[:-len(ending)]
    return thestring


def parse_cmd_line():
    project_file = ""
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hp:", ["project="])
    except getopt.GetoptError:
        print(CMD_HELP)
        sys.exit(2)
    for opt, arg in opts:
        print(opt + ": " + arg)
        if opt == '-h':
            print(CMD_HELP)
            sys.exit()
        elif opt in ("-p", "--project"):
            project_file = arg
    return project_file


def get_image_details():
    docker_image = os.getenv('CDB_DOCKER_IMAGE', "NO_DOCKER_IMAGE_FOUND")
    sha256_digest = os.getenv('CDB_ARTIFACT_SHA', None)
    if sha256_digest is None:
        client = docker.from_env()
        print("Inspecting docker image for sha256Digest")
        image = client.images.get(docker_image)
        sha256_digest = image.attrs["RepoDigests"][0].split(":")[1]
    else:
        print("Docker image digest found from environment variable")

    print("IMAGE DIGEST: " + str(image.attrs["RepoDigests"][0]))
    return docker_image, sha256_digest


def env_is_compliant():
    return os.getenv('CDB_IS_COMPLIANT', "FALSE") == "TRUE"


def is_compliant_suite(junit_xml):
    if junit_xml.failures != 0:
        return False, "Tests contain failures"
    if junit_xml.errors != 0:
        return False, "Tests contain errors"
    return True, "All tests passed"


def load_test_results(file_path):
    test_xml = JUnitXml.fromfile(file_path)
    return test_xml


def is_compliant_test_results(file_path):
    """
    This parses a junit xml file to determine if there are any errors or failures

    return:
        A tuple, with is_compliant, plus a message string
    """
    test_xml = load_test_results(file_path)
    if test_xml._tag == "testsuites":
        for suite in test_xml:
            suite_is_compliant, message = is_compliant_suite(suite)
            if not suite_is_compliant:
                return suite_is_compliant, message
        # every test suite passed, so return True
        return True, "All tests passed"
    if test_xml._tag == "testsuite":
        return is_compliant_suite(test_xml)
    return False, "Could not find test suite(s)"


def control_junit():
    print("Publish evidence to ComplianceDB")
    junit_results_path = "/data/junit/junit.xml"

    is_compliant, message = is_compliant_test_results(junit_results_path)
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "junit")
    description = "JUnit results xml verified by compliancedb/cdb_controls: " + message
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def put_evidence():
    print("Publish evidence to ComplianceDB")

    is_compliant = env_is_compliant()
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "EVIDENCE_TYPE_UNDEFINED")
    description = os.getenv('CDB_DESCRIPTION', "UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def send_evidence(evidence):
    print(evidence)
    project_file = parse_cmd_line()
    with open(project_file) as project_file_contents:
        _docker_image_unused, sha256_digest = get_image_details()
        api_token = os.getenv('CDB_API_TOKEN', 'NO_API_TOKEN_DEFINED')
        host = os.getenv('CDB_HOST', CDB_SERVER)
        add_evidence(api_token, host, project_file_contents, sha256_digest, evidence)


def build_evidence_dict(is_compliant, evidence_type, description, build_url):
    evidence = {"evidence_type": evidence_type, "contents": {
        "is_compliant": is_compliant,
        "url": "",
        "description": ""
    }}
    evidence["contents"]["description"] = description
    evidence["contents"]["url"] = build_url
    return evidence
