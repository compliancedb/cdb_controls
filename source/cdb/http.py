import json
import os

import requests as req
from requests.auth import HTTPBasicAuth
from cdb.http_retry import HttpRetry


def http_get_json(url, api_token):
    print("Getting json:")
    print("From this url: " + url)
    if in_dry_run():
        print("DRY RUN: Get not performed")
    else:
        auth = HTTPBasicAuth(api_token, 'unused')
        response = HttpRetry().get(url, auth=auth)
        print(response.text)
        return response.json()
    return None


def http_put_payload(url, payload, api_token):
    print("Putting this payload:")
    print(pretty_json(payload))
    print("To this url: " + url)
    if in_dry_run():
        print("DRY RUN: Put not sent")
    else:
        auth = HTTPBasicAuth(api_token, 'unused')
        headers = json_content_header()
        data = json.dumps(payload)
        response = HttpRetry().put(url, auth=auth, headers=headers, data=data)
        print(response.text)


def http_post_payload(url, payload, api_token):
    print("Posting this payload:")
    print(pretty_json(payload))
    print("To this url: " + url)
    if in_dry_run():
        print("DRY RUN: Post not sent")
    else:
        auth = HTTPBasicAuth(api_token, 'unused')
        headers = json_content_header()
        data = json.dumps(payload)
        response = HttpRetry().post(url, auth=auth, headers=headers, data=data)
        print(response.text)


def json_content_header():
    return {"Content-Type": "application/json"}


def pretty_json(payload):
    return json.dumps(payload, sort_keys=True, indent=4)


def in_dry_run():
    return os.getenv('CDB_DRY_RUN') == "TRUE" or os.getenv('MERKELY_DRY_RUN') == "TRUE"
