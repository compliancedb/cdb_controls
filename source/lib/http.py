from lib.http_retry import HttpRetry
import json
from requests.auth import HTTPBasicAuth


def http_get_json(url, api_token):
    auth = HTTPBasicAuth(api_token, 'unused')
    return HttpRetry().get(url, auth=auth)


def http_put_payload(url, payload, api_token):
    auth = HTTPBasicAuth(api_token, 'unused')
    headers = json_content_header()
    data = json.dumps(payload)
    return HttpRetry().put(url, auth=auth, headers=headers, data=data)


def http_post_payload(url, payload, api_token):
    auth = HTTPBasicAuth(api_token, 'unused')
    headers = json_content_header()
    data = json.dumps(payload)
    return HttpRetry().post(url, auth=auth, headers=headers, data=data)


def json_content_header():
    return {"Content-Type": "application/json"}
