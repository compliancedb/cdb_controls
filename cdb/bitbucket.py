import json
import os
import sys

import requests

from cdb.cdb_utils import build_evidence_dict, send_evidence


def put_bitbucket_pull_request(project_config_file):

    pull_requests = get_pull_request_for_current_commit()

    print("Found pull request info: " + str(pull_requests))
    evidence = build_evidence_dict(
        is_compliant=True,
        evidence_type="pull_request",
        description="Bitbucket pull request",
        build_url=pull_requests[0]["pullRequestURL"])
    evidence["contents"]["source"] = pull_requests
    send_evidence(project_config_file, evidence)


def get_pull_request_for_current_commit():
    workspace = os.getenv('BITBUCKET_WORKSPACE', None)
    repository = os.getenv('BITBUCKET_REPO_SLUG', None)
    commit = os.getenv('BITBUCKET_COMMIT', None)
    user = os.getenv('BITBUCKET_USER', None)
    password = os.getenv('BITBUCKET_PWD', None)
    pull_requests = get_pull_requests_from_bitbucket_api(
        workspace=workspace, repository=repository, commit=commit, username=user, password=password)
    return pull_requests


def get_pull_requests_from_bitbucket_api(workspace, repository, commit, username, password):
    pull_requests_evidence = []

    url = f"https://api.bitbucket.org/2.0/repositories/{workspace}/{repository}/commit/{commit}/pullrequests"
    r = requests.get(url, auth=(username, password))
    if r.status_code == 200:
        pull_requests_json = json.loads(r.text)
        pull_requests = pull_requests_json["values"]
        for pr in pull_requests:
            pr_evidence = {}

            pr_html_url = pr['links']['html']['href']
            pr_api_url = pr['links']['self']['href']

            pr_evidence['pullRequestMergeCommit'] = commit
            pr_evidence['pullRequestURL'] = pr_html_url

            pr_evidence = get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password)

            pull_requests_evidence.append(pr_evidence)

    elif r.status_code == 202:
        print("Repository pull requests are still being indexed, please retry.")
        sys.exit(1)
    elif r.status_code == 404:
        print("Repository does not exists or pull requests are not indexed. Please make sure Pull Request Commit Links app is installed")
        sys.exit(2)
    else:
        print(f"Exception occured in fetching pull requests. Http return code is {r.status_code}")
        sys.exit(3)

    return pull_requests_evidence


def get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password):
    r = requests.get(pr_api_url, auth=(username, password))
    if r.status_code == 200:
        pr_json = json.loads(r.text)
        pr_evidence['pullRequestState'] = pr_json['state']
        participants = pr_json['participants']
        approvers = ""
        if len(participants) > 0:
            for participant in participants:
                if participant['approved']:
                    approvers = approvers + participant['user']['display_name'] + ","
            approvers = approvers[:-1]
            pr_evidence['approvers'] = approvers
        else:
            print("No approvers found")
    else:
        print("Error occured in fetching pull request details. Please review repository permissions.")
    return pr_evidence





