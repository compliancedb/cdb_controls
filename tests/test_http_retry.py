from cdb.http import http_post_payload, http_put_payload, http_get_json
import cdb.http_retry as http_retry

from pytest import raises
from tests.utils import verify_approval, stub_http_503, retry_backoff_factor
import responses

MAX_RETRY_COUNT = 5


def test_total_retry_sleep_time_is_about_30_seconds():
    assert http_retry.HttpRetry().total_sleep_time() == 31  # 1+2+4+8+16


@responses.activate
def test_503_post_retries_5_times_then_raises_RetryError(capsys):
    url, payload, api_token = stub_http_503('POST', 1+MAX_RETRY_COUNT)

    with retry_backoff_factor(0.001), raises(http_retry.Error) as exc_info:
        http_post_payload(url, payload, api_token)

    assert exc_info.value.url() == url
    assert len(responses.calls) == 1+MAX_RETRY_COUNT
    verify_approval(capsys)


@responses.activate
def test_503_put_retries_5_times_then_raises_RetryError(capsys):
    url, payload, api_token = stub_http_503('PUT', 1+MAX_RETRY_COUNT)

    with retry_backoff_factor(0.001), raises(http_retry.Error) as exc_info:
        http_put_payload(url, payload, api_token)

    assert exc_info.value.url() == url
    assert len(responses.calls) == 1+MAX_RETRY_COUNT
    verify_approval(capsys)


@responses.activate
def test_503_get_retries_5_times_then_raises_RetryError(capsys):
    url, _, api_token = stub_http_503('GET', 1+MAX_RETRY_COUNT)

    with retry_backoff_factor(0.001), raises(http_retry.Error) as exc_info:
        http_get_json(url, api_token)

    assert exc_info.value.url() == url
    assert len(responses.calls) == 1+MAX_RETRY_COUNT
    verify_approval(capsys)


@responses.activate
def test_post_stops_retrying_when_non_503_and_returns_None(capsys):
    url, payload, api_token = stub_http_503('POST', 1+1)

    with retry_backoff_factor(0.001):
        response = http_post_payload(url, payload, api_token)

    assert response is None
    assert len(responses.calls) == 1+1+1
    verify_approval(capsys)


@responses.activate
def test_put_stops_retrying_when_non_503_and_returns_None(capsys):
    url, payload, api_token = stub_http_503('PUT', 1+1)

    with retry_backoff_factor(0.001):
        response = http_put_payload(url, payload, api_token)

    assert response is None
    assert len(responses.calls) == 1+1+1
    verify_approval(capsys)


@responses.activate
def test_get_stops_retrying_when_non_503_and_returns_response_json(capsys):
    url, _, api_token = stub_http_503('GET', 1+1)

    with retry_backoff_factor(0.001):
        response_json = http_get_json(url, api_token)

    assert response_json == {"success": 42}
    assert len(responses.calls) == 1+1+1
    verify_approval(capsys)
