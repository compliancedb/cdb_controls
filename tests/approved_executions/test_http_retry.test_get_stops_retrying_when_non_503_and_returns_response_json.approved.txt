Getting json:
From this url: https://test.compliancedb.com/api/v1/get/
Response.status=503, retrying in 0.001 seconds...
Retry 1/5: response.status=503, retrying in 0.002 seconds...
Retry 2/5: response.status=200
{"success": 42}
